### PR TITLE
proto: utility functions ToProto and FromProto

### PR DESCRIPTION
**Pull request by fedekunze**
## Description

Wrote more util functions for the following types: `BlockID`, `Proposal`, `Consensus`, `BitArray`, `Commit` and `PartSetHeaders`.

______

For contributor use:

- [x] Wrote tests
- [x] Updated CHANGELOG_PENDING.md
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: types/block.go [hunk 1]
```diff
@@ -804,21 +815,29 @@ func (data *Data) StringIndented(indent string) string {
 		indent, data.hash)
 }
 
+// ToProto converts Data to protobuf
 func (data *Data) ToProto() *tmproto.Data {
+	if data == nil {
```
Comment (fedekunze) [inline types/block.go:820]: @melekes @tess do these safety checks make sense to you? otherwise, the functions may panic

Comment (melekes) [reply types/block.go:820]: Fine by me if data indeed can be nil at any point during the lifetime.

I'd prefer for us to use https://en.wikipedia.org/wiki/Null_Object_pattern, but it may not be possible here.

File: types/header.go [hunk 2]
```diff
@@ -179,15 +179,18 @@ func (h Header) ValidateBasic() error {
...
 func (h *Header) ToProto() *tmproto.Header {
-	if h == nil {
-		return nil
-	}
-	return &tmproto.Header{
-		Version:            h.Version.ToProto(),
-		ChainID:            h.ChainID,
-		Height:             h.Height,
-		Time:               h.Time,
-		LastBlockID:        h.LastBlockID.ToProto(),
```
Comment (fedekunze) [inline types/header.go]: can you undo this change? @marbar3778

Comment (marbar3778) [reply types/header.go]: the toProto?

File: libs/bits/bit_array_test.go [hunk 3]
```diff
@@ -265,3 +265,25 @@ func TestJSONMarshalUnmarshal(t *testing.T) {
...
+		expPass bool
+	}{
+		{"success empty", &BitArray{}, true},
+		{"success", NewBitArray(1), true},
+		{"success", NewBitArray(2), true},
+		{"negative", NewBitArray(-1), false},
+	}
+	for _, tc := range testCases {
+		protoBA := tc.bA1.ToProto()
+		fmt.Println(protoBA)
```
Comment (fedekunze) [inline libs/bits/bit_array_test.go:282]:
```suggestion
```

File: libs/bits/bit_array_test.go [hunk 4]
```diff
@@ -265,3 +265,25 @@ func TestJSONMarshalUnmarshal(t *testing.T) {
...
+		ba := new(BitArray)
+		ba.FromProto(protoBA)
+		if tc.expPass {
```
Comment (fedekunze) [inline libs/bits/bit_array_test.go:285]: else?

File: types/block_test.go [hunk 5]
```diff
@@ -599,3 +599,187 @@ func TestBlockIDValidateBasic(t *testing.T) {
...
+	}{
+		{"success", data, true},
+		{"success data2", data2, true},
+	}
+	for _, tc := range testCases {
+		protoData := tc.data1.ToProto()
+		d := new(Data)
+		err := d.FromProto(protoData)
+		if tc.expPass {
+			require.EqualValues(t, tc.data1, d, tc.msg)
```
Comment (fedekunze) [inline types/block_test.go:661]:
```suggestion
			require.NoError(t, err, tc.msg)
			require.EqualValues(t, tc.data1, d, tc.msg)
```
